docstore-metrics/src/main/java/org/hypertrace/core/serviceframework/docstore/metrics/model/GaugeBuilderProvider.java

```diff
@@ -0,0 +1,29 @@
+package org.hypertrace.core.serviceframework.docstore.metrics.model;
+
+import static org.hypertrace.core.serviceframework.metrics.PlatformMetricsRegistry.toIterable;
+
+import io.micrometer.core.instrument.Gauge;
+import java.time.Duration;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import lombok.Builder;
+
+@Builder
+public class GaugeBuilderProvider<T extends Number> {
+  private final String name;
+  private final Map<String, String> tags;
+  private final T number;
+  private final Duration duration;
+  private final ScheduledExecutorService executor;
+
+  public Gauge.Builder<T> get() {
+    // Not creating a strong reference which indefinitely avoids the number from being GCed
```

Comment:
I'm confused about this - if I'm reading this right, it seems like the issue is we're creating a new gauge every single report because we built an API that takes a constant value (number) and wraps it in a function. We should fix the metric registry (without breaking it) to allow building a gauge a single time with a number supplier rather than a constant number.

Neither that, nor what we have here would fix the issue you're seeing however. I think the right solution here probably is using a multigauge - https://docs.micrometer.io/micrometer/reference/concepts/gauges.html#_multi_gauge

Reply:
Makes sense. Updated.